*: break dependency loop between `sqlutil` and `eval`

ZhouXing19 opened the pull request:
We'd like to introduce 

```
	WithInternalExecutor(
		ctx context.Context,
		run func(ctx context.Context, txn *kv.Txn, ie sqlutil.InternalExecutor) error,
	) error
```

into the `eval.Planner` interface, but encountered the following loops:

```
.-> //pkg/sql/sqlutil:sqlutil (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog/colinfo:colinfo (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/sem/eval:eval (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
`-- //pkg/sql/sqlutil:sqlutil (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d

.-> //pkg/sql/sqlutil:sqlutil (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog:catalog (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/sem/eval:eval (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
`-- //pkg/sql/sqlutil:sqlutil (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
``` 

This PR is to break these 2 loops. 

----
More context:

We'd like to refactor this usage of internal executor by binding it to the txn
https://github.com/cockroachdb/cockroach/blob/703aedb45ebe15cac630bf7f85950db344ea9776/pkg/jobs/jobs.go#L786-L791
The difficulty I find is that if we trace its usage, the function is called in `Registry.LoadJobWithTxn()`, which has [a usage](https://github.com/cockroachdb/cockroach/blob/703aedb45ebe15cac630bf7f85950db344ea9776/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L66) with txn eventually from evalCtx.txn . So we have to have an internal executor bound to evalCtx.txn. 

Epic: CRDB-19135

Release note: None

ZhouXing19 committed:
sqlutil: break dependency on catalog

Break `sqlutil`'s' direct dependency on `catalog` in

```
.-> //pkg/sql/sqlutil:sqlutil (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog:catalog (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/sem/eval:eval (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
`-- //pkg/sql/sqlutil:sqlutil (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
```

Added a new pkg `descriptorlist` for an interface `CatalogDescriptorList` that
corresponds to `catalog.Descriptors`. It is used in `sqltuil.InternalExecutor`'s
function, so that the `sqlutil` can break dependency on `catalog`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/90781)
<!-- Reviewable:end -->

ZhouXing19 committed:
eval: add WithInternalExecutor() to eval.Planner

This commit is to add WithInternalExecutor() function to the `eval.Planner`
interface.

Release note: None

ZhouXing19 commented:
Hmm, maybe this won't work. I was trying to break "`catalog` imports `eval`", but i just noticed that the other way is [prohibited already](https://github.com/cockroachdb/cockroach/blob/9897efac2d0b18b5921028ac450e0929df16cb5b/pkg/sql/sem/eval/BUILD.bazel#L143-L148). I.e., what I'm having now, as below, is not allowed
<img width="489" alt="Screen Shot 2022-10-27 at 15 57 19" src="https://user-images.githubusercontent.com/53610260/198386188-876136f9-9a2e-4e26-81af-909361cd2abb.png">

ZhouXing19 committed:
colinfo: break dependency on catalog

This commit is to break the packages `colinfo`'s dependency on `catalog`,
as in this loop:

```
.-> //pkg/sql/sqlutil:sqlutil (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog/colinfo:colinfo (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog:catalog (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/sem/eval:eval (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
`-- //pkg/sql/sqlutil:sqlutil (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
```

We created a new package `pkg/sql/catalog/catalogtableelements` to store an
interface `catalogtableelements.Columns` which corresponds to `catalog.Columns`.

We also moved functions from `catalog` to some other pkgs.

Release note: None

ZhouXing19 committed:
colinfo: break dependency on eval

Release note: None

ajwerner reviewed (commented):
I think this PR is missing motivation on why you want to expose access to creating a new transaction in the `EvalPlanner`. That's a very dangerous idea because it's so easy to create deadlocks. I'm pretty opposed to adding such power to the already worst interface in the code bade. 

> Since it doesn't help to break "catalog imports eval" , maybe I should try breaking "colinfo imports catalog"?

I'd love for us to break this dependency. It's a very bad one. It'll take some work. https://github.com/cockroachdb/cockroach/issues/88400

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90781#-:-NFffPpL7fqY4_az1Vs5:b-8058hu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rhu713)



<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
> I think this PR is missing motivation on why you want to expose access to creating a new transaction in the EvalPlanner.

It's because in some places (e.g. `replication_builtins.go`), the internal executor is bound to `eval.Txn` to run a query. For example:
https://github.com/cockroachdb/cockroach/blob/b70d9ddfbce8e08430ea36ebad3ea5f51952f831/pkg/sql/sem/builtins/replication_builtins.go#L99

I also put an example in the PR description.

This led me to think that we may want to have an internal executor bound to `eval.Planner`, if people wants to use it with a txn. 

> That's a very dangerous idea because it's so easy to create deadlocks.

Could you say more on this? I'm not opposing but just curious about more details.

ajwerner commented:
> It's because in some places (e.g. replication_builtins.go), the internal executor is bound to eval.Txn to run a query. For example:

I think that's a pretty bad example of dependencies leaking through eval where instead we ought to rework the interfaces to have a more narrow method in builtins. Builtins shouldn't be plumbing big dependencies into packages. It should be invoking narrow interfaces which were set up above it. Any time you're plumbing dependencies through to some sort of constructor in builtins, alarm bells should be going off.

> Could you say more on this? I'm not opposing but just curious about more details.

If the outer transaction holds a lock on a row, and then during execution a new transaction is created which attempts to interact with that locked row, there will be a deadlock cycle but the system will never be able to detect it. We absolutely must not make it easy to create transactions underneath evaluation. There be dragons that way.

ajwerner commented:
We should try hard to eliminate `eval.Txn` and we should try hard to eliminate any use of a Txn anywhere in the eval interfaces.

ajwerner commented:
We should rework the streaming/api.go and the related things in the eval context. I think the streaming APIs are pretty bad. I don't get why you'd want to pass an eval context to construct the `Manager` types and then also pass both that thing and the `Txn` to the methods. That seems silly to me. I think what I'd push for is a way to have `eval` define interface like:

```go

type StreamIngestionManager interface {
	// CompleteStreamIngestion signals a running stream ingestion job to complete on the consumer side.
	CompleteStreamIngestion(
		ctx context.Context,
		ingestionJobID jobspb.JobID,
		cutoverTimestamp hlc.Timestamp,
	) error

	// GetStreamIngestionStats gets a statistics summary for a stream ingestion job.
	GetStreamIngestionStats(
		ctx context.Context,
		ingestionJobID jobspb.JobID,
	) (*streampb.StreamIngestionStats, error)
}

type ReplicationStreamManager interface {
	// StartReplicationStream starts a stream replication job for the specified tenant on the producer side.
	StartReplicationStream(
		ctx context.Context,
		tenantID uint64,
	) (streaming.StreamID, error)

	// HeartbeatReplicationStream sends a heartbeat to the replication stream producer, indicating
	// consumer has consumed until the given 'frontier' timestamp. This updates the producer job
	// progress and extends its life, and the new producer progress will be returned.
	// If 'frontier' is hlc.MaxTimestamp, returns the producer progress without updating it.
	HeartbeatReplicationStream(
		ctx context.Context,
		streamID StreamID,
		frontier hlc.Timestamp,
	) (streampb.StreamReplicationStatus, error)

	// StreamPartition starts streaming replication on the producer side for the partition specified
	// by opaqueSpec which contains serialized streampb.StreamPartitionSpec protocol message and
	// returns a value generator which yields events for the specified partition.
	StreamPartition(
		ctx context.Context,
		streamID StreamID,
		opaqueSpec []byte,
	) (ValueGenerator, error)

	// GetReplicationStreamSpec gets a stream replication spec on the producer side.
	GetReplicationStreamSpec(
		ctx context.Context,
		streamID StreamID,
	) (*streampb.ReplicationStreamSpec, error)

	// CompleteReplicationStream completes a replication stream job on the producer side.
	// 'successfulIngestion' indicates whether the stream ingestion finished successfully and
	// determines the fate of the producer job, succeeded or canceled.
	CompleteReplicationStream(
		ctx context.Context,
		streamID StreamID,
		successfulIngestion bool,
	) error
}
```
Or something like that and then it's fine to implement these interfaces on `sql.Planner` or have constructors for these interfaces on sql.Planner. Nevertheless, we ought to inject those via different fields on the `eval.Context`.

ZhouXing19 commented:
> Or something like that and then it's fine to implement these interfaces on sql.Planner or have constructors for these interfaces on sql.Planner.

Thanks! I just had https://github.com/cockroachdb/cockroach/pull/90964 up for this.

ZhouXing19 commented:
We may want to hold this till https://github.com/cockroachdb/cockroach/pull/90964 is finished. Converting it back to draft.
